### ISSUE TITLE
Proposal for the baseline images problem

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#16447]
This is a proposal to help with the difficulty in adding/modifying tests which require a baseline image.  As a brief reminder, baseline images are problematic because
- they cause the repo size to grow rather quickly (matplotlib's .git is now ~400Mb, compared e.g. to numpy's which is 72Mb), even though we've been exceedingly careful in recent years to avoid adding more baseline images (the problems with the merge of #8896 got me to think about this again).
- they force us to pin to a somewhat old version of freetype, because nearly *every* release of freetype causes tiny rasterization changes that would entail regenerating all baseline images (and thus cause even more repo size growth).

The idea is to actually *not* store baseline images in the repo at all.
1) When generating a sdist or wheels for a new release, we do generate baseline images (from the tests) and store them in the released sdist or wheel.  The validity of the baseline images is guaranteed by what comes below.
2) During normal development, the baseline images should not change.  If someone wants to run the test suite from a git checkout, matplotlib would first generate a set of "known" baseline images either by grabbing the last release, or just installing the last tagged release into a temporary venv and running the test suite there (of course, they would be cached into ~/.cache/matplotlib so this is only done once -- this caching also needs to be done so that CI time doesn't double...).
3) If a baseline image needs to change (we should still try to keep this as rare as possible), the *commit message* needs to include a special string e.g. `[baseline-change]: path/to/image`.  The process described in 2) would also consult the git log to know if there's any images that need to be gen'd in such a way.  Likewise, if a baseline image needs to be added for a new test, the commit message needs to be tagged accordingly.  Possibly, we could also require that the filename also be changed in that case so that even if you're not in a git repo, you at least don't end up using the wrong baseline image, and instead just get a skip.

Note that this means that if you're using a non-git-repo, non-sdist/wheel, you cannot run the full test suite anymore, because you at least need to know what was the last release, and if any image changed since then, you'd like to know when (however, if you want to write a patch and check that it's good, this still works because you can compare the baselines with what was present before your patch).  OTOH distro/conda packagers, which *should* be working from sdists, would still be able to run the test suite fine, because the sdists/wheels will include baseline images per 1).

There's a bit of machinery involved in making this work, but I don't think it's overly hard to implement.  Thoughts?

**Comment by QuLogic** [comment]
At least as far as getting new baseline images with other FreeType, you can grab them from https://github.com/QuLogic/mpl-images.

I have not yet completely thought about the rest of the proposal, but I suspect if caching is not sufficient, it might be possible to do the remaining work in a GitHub Action.

**Comment by tacaswell** [comment]
This is the best proposal for handling this problem yet!

 In summary:
 - on release, generate the 'blessed' version of the test images for that tag (these could even be thrown onto github as artifacts with the release)
 - on a dev machine, generate and/or download and cache the test images running against the last released version and/or current master
 - some logic to flag what images have been changed from last version to the current branch tip and do something about them
 - for CI we can rely on CI cache or stand up a store on github/DO/AWS/... managing that cache can be done via github actions

As part of this we should publish `mpl-test-data` as a wheel separately.  

If we have the tools for CI to automatically grab the latest baseline images based on the current commit, might as well have that be the recommended way for devs to get the test images as well (with the generate-in-place being an option if you have local CPU/time, but poor internet)?

One thing we might lose with this is that we currently use the exactly same version of the images on all OS, all python versions, all machines etc.  I think we can maintain this, but it is a bit trickier to be 100% sure that we have it right (as opposed to "git says it is these bits")

A thing that this would give us is a nice way to generate forward and backwards compatibility reports (run the next tag with the baseline images from the previous tag, list what fails).

This would also get us around the problem that some of our latex baselines are different with newer versions of latex.

Are we re-inventing git LFS?

**Comment by anntzer** [comment]
> Are we re-inventing git LFS?

If you add the machinery to grab the latest test images, perhaps; but the intent was really that CI uses CI cache to roll over one run's generated test images as baselines for the next run, and that devs do the same.  Which is really the fundamental idea here: instead of comparing against a fixed baseline, compare a commit to the next (with some additional shortcuts based on reference points at releases, etc.).

**Comment by jklymak** [comment]
I'm not fully following this, but it sounds painful for contributors.  If we are going to be painful to contributors, it seems to me that just having a separate images repository on github would be simpler.  Someone would have to tag it to releases, so maybe that makes the release work harder?  But then folks could just do shallow fetches if they do not need the full history of the images.  But perhaps I'm missing something, or maybe thats more complicated.

**Comment by anntzer** [comment]
This should transparent for contributors (except that their first test run is going to be a bit slower *if some baseline images changed since the last release*, given that mpl will automatically rebuild itself at the correct commit to generate the corresponding baseline).

**Comment by SidharthBansal** [comment]
Hi, I am interested in this issue. I am new to the organisation. May I know how shall I proceed? I am preparing the proposal and submitting it asap! Will be working more on it too!

**Comment by SaraLatif99** [comment]
Hi, I am interested, I look forward to woorking on it.

**Comment by anntzer** [comment]
See https://github.com/numfocus/gsoc.

**Comment by SidharthBansal** [comment]
Right now I am installing the matplotlib

**Comment by SidharthBansal** [comment]
Installed it. Thanks. Looked at the above link @anntzer thanks

**Comment by laksh9950** [comment]
I am interested in this issue . This is the kind of problem i would like to work on my summer time . As i Know testing is an important part of a development process . From what i understand we can actually store the binary of baseline images and their format in the release . I need to work a lot more than that. @anntzer  please  guide me
Thanks

**Comment by anntzer** [comment]
See some suggestions at https://discourse.matplotlib.org/t/gsoc-due-march-31-18-00utc/20975/10?u=anntzer.lee as well.

**Comment by laksh9950** [comment]
> See some suggestions at https://discourse.matplotlib.org/t/gsoc-due-march-31-18-00utc/20975/10?u=anntzer.lee as well.

yeah i have looked at it . i have on python testing in one of internships but i have to get comfortable with the codebase and then i can do a pull request .
I need some help in  my proposal .  can you please help @anntzer

**Comment by SidharthBansal** [comment]
> See some suggestions at https://discourse.matplotlib.org/t/gsoc-due-march-31-18-00utc/20975/10?u=anntzer.lee as well.

Okay, I am looking towards them. Thanks

**Comment by SidharthBansal** [comment]
Individual changes listed above like generation of baseline images from new release in wheels, modification of baseline images logics, CI changes are quite vast. So, I think we should tackle them in different PRS after breaking the issue. This will also ensure modularity. But, independent pull requests merged into master, can leave master in an inconsistent state.
So, we should create a beta branch and get them merged into beta branch. Once the beta branch is fully completed, then we can move the changes to the master. What do you suggest?

**Comment by anntzer** [comment]
We typically prefer merging things in a piecemeal fashion but if that's too tricky (I realize it's going to be a lot of work, that's why it's a GSOC :-)) then I guess putting everything on a branch may be an option; thoughs @tacaswell @QuLogic or anyone else?

**Comment by laksh9950** [comment]
Actually the problem is not to include the baseline images In repo itself
.developer working on the project get the baseline images using previous
versions as the baseline images can be cached in the memory. I would to get
more thoughts on that . Can we include baseline images with a pip package .
@anntzer

**Comment by anntzer** [comment]
> Can we include baseline images with a pip package.

Yes, certainly, that's part of the plan too.  That's in fact necessary so that people who test just based on the pip package (e.g., Linux distro packagers) can still run the tests.

**Comment by laksh9950** [comment]
> > Can we include baseline images with a pip package.
> 
> Yes, certainly, that's part of the plan too. That's in fact necessary so that people who test just based on the pip package (e.g., Linux distro packagers) can still run the tests.

Yes , I am looking at the codebase and looking for the ways to include baseline images in the pip package .

**Comment by SidharthBansal** [comment]
Few questions:

**1. Use Case 1**: Suppose a user of matplotlib(not a contributor) is A. So, I think A will never get involved in testing. So, in the entire lifespan of A(assuming he never runs pytest), he will never have any baseline images on their computer, nor his fork repository. Please correct me if I am wrong.
There is no use for A, to grab baseline images as they are not involved in testing.

**2. Use Case 2:** Developer(maybe also a user of mpl) runs only one test (out of entire Test Suite) for a particular feature testing. Feature involves only 2 baseline images B1 and B2. Entire set of baseline images B. So, will this developer download B or (B1+B2)?

_Use Case solution 2A:_ Download B-> First time, this will take a lot of time for downloading entire set of baseline images from the last release(having all baseline images). Later on, time will be reduced for testing. 

_Use Case solution 2B:_ Download B1+B2: Faster download of B1+B2 but a little bit of overhead each time to connect to the last release to grab baseline images. Also, there will need to check which baseline images are present on the developer's desktop. Only images which are not present on the developer's desktop needs to be downloaded from the last release. So, logic regarding the comparison between the last release and the current developer's master branch needs to be created.

**Comment by anntzer** [comment]
> Use Case 1: Suppose a user of matplotlib(not a contributor) is A. So, I think A will never get involved in testing. So, in the entire lifespan of A(assuming he never runs pytest), he will never have any baseline images on their computer, nor his fork repository. Please correct me if I am wrong.
There is no use for A, to grab baseline images as they are not involved in testing.

Yes.

> Use Case 2: Developer(maybe also a user of mpl) runs only one test (out of entire Test Suite) for a particular feature testing. The feature involves only 2 baseline images B1 and B2. The entire set of baseline images B. So, will this developer download B or (B1+B2)?

The plan is to download everything.

> Use Case solution 2A: Download B-> First time, this will take a lot of time for downloading the entire set of baseline images from the last release(having all baseline images). Later on, time will be reduced for testing.

Yes, but the total download is not worse than currently, because currently you *always* download all baseline images anyways.

> Use Case solution 2B: Download B1+B2: Faster download of B1+B2 but a little bit of overhead each time to connect to the last release to grab baseline images. Also, there will need to check which baseline images are present on the developer's desktop. Only images which are not present on the developer's desktop needs to be downloaded from the last release. So, logic regarding the comparison between the last release and the current developer's master branch needs to be created. Another subcase could be, the developer deleted some of the previously downloaded baseline images. So, while checking the last release, developer should be able to download those images which no longer exists on his master.

> Considering coupling and fact that one test may break non-correspondent test(by indirect calling), it is better to fetch all baseline images at once. Also, most of the times developer run full test suite. Also, increasing wait time each time for the B1+B2 is irritating for the developer. I prefer 2A.

It's also more infrastructure to set up to be able to do piecemean downbloads.

> Use Case 3:Deletion of baseline images: I observed that deletion of baseline image is not mentioned above. Suppose a plotting feature is deprecated. So, in later releases, there will need to delete the baseline images related to the deprecated images. Also, deletion of baseline images (if developed as a reusable submodule/method) can be used for base

That seems fine?

-----

One important case that needs to be mentioned is if a baseline image *changes* but there's no release yet with that change.  In that case the developer cannot download that baseline image.  Instead (this is one important point of the proposal), this change should be marked in some way in the git history, the testing machinery should detect that, set up a temporary environment with the last "known good" version of Matplotlib, run the test suite locally there to generate the corresponding baseline image, cache it locally (so that we don't have to redo this every time), and then use that as baseline image to run the test on master.

**Comment by SidharthBansal** [comment]
> Yes, but the total download is not worse than currently, because currently you always download all baseline images anyways.

That is the reason I asked whether we need 2A or 2B. So, aiming at entire baseline image set will also reduce the code complexity. :)

> That seems fine?

Suppose we modified baseline images subset B = {B1, B2...., Bn}. So, we deleted things from release Rx and current release is Ry such that they both may not be concurrent(y-x >= 1). So, we will be adding the modified B to Ry so that these modified features can be continuously integrated and tested by other developers on current Ry. Other plan will be to include it in Ry+1. I am in favor of Ry. Both of the scenarios will change the release version number from which baseline images are downloaded. What do you think?

**Comment by SidharthBansal** [comment]
What does ` because the sdists/wheels will include baseline images per 1` mean in issue body?

**Comment by SidharthBansal** [comment]
When are we aiming at merging https://github.com/matplotlib/matplotlib/pull/11732 ? Any plans?

**Comment by SidharthBansal** [comment]
> Note that this means that if you're using a non-git-repo, non-sdist/wheel, you cannot run the full test suite anymore, because you at least need to know what was the last release, and if any image changed since then, you'd like to know when (however, if you want to write a patch and check that it's good, this still works because you can compare the baselines with what was present before your patch).

My opinion is #11732 can simplify #16447 to a large extend. If we are not maintaining separate test and test-data wheels then each time we need to detect the latest set of baseline images belong to which release. If separation is maintained, then logic of comparision will be simplified to last test-wheel release instead of entire git history.

What do you mean by non-git repo above? Are you considering the future scenario of shifting matplotlib to gitlab etc. I think I am missing something here.

**Comment by SidharthBansal** [comment]
@QuLogic helped me in knowing CI used here on gitter by commenting:
> It's currently a mix of several CI systems: Travis (Linux/macOS), AppVeyor (Windows), Azure (all platforms), CircleCI (docs-build)

I just want to know are we using Continuous Delivery/Continuous Deployement? I have not seen Jenkins here so far. So, was curious to know how we deploy things here.

**Comment by anntzer** [comment]
> Suppose we modified baseline images subset B = {B1, B2...., Bn}. So, we deleted things from release Rx and current release is Ry such that they both may not be concurrent(y-x >= 1). So, we will be adding the modified B to Ry so that these modified features can be continuously integrated and tested by other developers on current Ry. Other plan will be to include it in Ry+1. I am in favor of Ry. Both of the scenarios will change the release version number from which baseline images are downloaded. What do you think?

I think we can consider that Matplotlib only ever has one "active" branch from the PoV of baseline images.  (The last long-lived branch was 2.2.x for Py2 support, but that has been closed now.)

So for example right now 3.2.x is the last release and people are working on the next release which is 3.3; for simplification let's say that we will *never* make changes to baseline images on the 3.2.x branch as that would count as API changes; the only branch on which baseline images changes will occur is 3.3.  (Again, we may want to relax this restriction later but I don't think it's unreasonable.)

So on PyPI there's the set of 3.2.0 baseline images available ("matplotlib-tests", as discussed below); the main remaining problem is for developers who are currently working on 3.3; they may want to run tests locally before 3.3 is released (once 3.3 is released we simultaneously release "matplotlib-tests 3.3" and again everyone has access to it and can download it).  (Note that developers is not just the "experienced" group, there's also the "drive-by" contributors who may just want to put in a PR; it should remain as simple as possible for them to just clone the repo, write their feature, and run the test suite locally.)  If between 3.2.0 and the current master (somewhere on the 3.3 branch before the 3.3.0 release) no images have changed, then it's simple, the test machinery can just download "matplotlib-tests-3.2.0" and use that.  The tricky part is if some images *have* changed in between (now the test machinery needs to generate baseline images using a previous commit, e.g. the one just after the image has changed, where the image is known to be good).

(As a side point, I think another simplification for now may be to say that baseline image never "change", they just get deleted and new ones added -- it's not too onerous to add a counter to the filename of images we want to modify.)

> What does "because the sdists/wheels will include baseline images per 1 mean in issue body"?

The idea was that if someone downloads matplotlib from PyPI (either as a .tar.gz or as a .whl), they should get something self-contained from which it should be possible to run the test suite without any further internet access or git repo or whatnot -- as they do now.  Although this idea should be nuanced by the possibility that the test data actually be in a separate PyPI package ("matplotlib-tests"), such as the one proposed by #11732.

> When are we aiming at merging #11732 ? Any plans?

IIRC the main problem there was that I wanted to ensure that the "matplotlib-notestdata" PyPI package (which should be just called "matplotlib" but which I'll call "matplotlib-notestdata" here just to avoid confusion) should *never* include the test data (otherwise, someone who installs both "matplotlib-notestdata" and "matplotlib-tests" will have the second package overwrite files in the first, likely causing all sorts of confusion -- https://github.com/pypa/pip/issues/3868 and linked issues.)

Also I don't exactly remember how #11732 interacts with editable installs (please read about them if you don't know them; our developer guide explicitly suggests using editable installs for developing Matplotlib).  Probably fine, but worth checking too.

> My opinion is #11732 can simplify #16447 to a large extend. If we are not maintaining separate test and test-data wheels then each time we need to detect the latest set of baseline images belong to which release. If separation is maintained, then logic of comparison will be simplified to last test-wheel release instead of entire git history. What do you think?

Yes, I think getting #11732 to work will help a lot.  It's just a matter of ironing out the edges.

> What do you mean by non-git repo above? Are you considering the future scenario of shifting matplotlib to gitlab etc. I think I am missing something here.

Copying my reply on gitter, but feel free to ask for clarifications:
We only have a git repo; the non-git repo case is that people may just have a tarball (or rather, a sdist, in python terminology, which is essentially a zip with some metadata), and still want to run tests from that; this is quite common e.g. for linux packagers.

**Comment by SidharthBansal** [comment]
> the main remaining problem is for developers who are currently working on 3.3; they may want to run tests locally before 3.3 is released. If between 3.2.0 and the current master (somewhere on the 3.3 branch before the 3.3.0 release) no images have changed, then it's simple, the test machinery can just download "matplotlib-tests-3.2.0" and use that.

There is a caveat here also I guess. Right now, version 3.2 is used by developers which have test data. And we will be having mpl-test and mpl-no-test after #11732 is complete. So, if there are no additions nor deletions(using terminology of the addition or deletion from now onwards instead of the modification term), then there is no need to download the test-data(as developers have test data already, (in test folder)). The download will only happen if mpl-test in release 3.3 is different from test folder in release 3.2. What do you think?

Also, for download there is no need to download the whole `mpl-tests`, can we compare things online somehow and download only the files which are deleted/added? Or are we planning to download the whole package and then run adjusting? If we will download mpl-test every time, I think it will be time consuming.

**Comment by anntzer** [comment]
> There is a caveat here also I guess. Right now, version 3.2 is used by developers which have test data. And we will be having mpl-test and mpl-no-test after #11732 is complete. So, if there are no additions nor deletions(using terminology of the addition or deletion from now onwards instead of the modification term), then there is no need to download the test-data(as developers have test data already, (in test folder)). The download will only happen if mpl-test in release 3.3 is different from test folder in release 3.2. What do you think?

Yes.

> Also, for download there is no need to download the whole mpl-tests, can we compare things online somehow and download only the files which are deleted/added? Or are we planning to download the whole package and then adjusting? If we will download mpl-test every time, I think it will be time consuming.

I think best would be to just download it once and cache it, so that you can run tests offline after that.  (One set of baseline images is not *that* big, a couple of dozen Mb, so caching locally is fine, what takes a lot of space is having all versions of all baseline images ever.)

**Comment by SidharthBansal** [comment]
> IIRC the main problem there was that I wanted to ensure that the "matplotlib-notestdata" PyPI package (which should be just called "matplotlib" but which I'll call "matplotlib-notestdata" here just to avoid confusion) should never include the test data (otherwise, someone who installs both "matplotlib-notestdata" and "matplotlib-tests" will have the second package overwrite files in the first, likely causing all sorts of confusion -- pypa/pip#3868 and linked issues.)

So after merging #11732, release 3.3 of matplotlib(which doesn't contain tests) will override the old package release 3.2 matplotlib(which contains tests). So, this upgrade will decrease storage for users(not developers) of mpl as now tests will be deleted from their computers. Is it right?
I believe users will be using commands like `pip install --upgrade matplotlib` for upgrades.

For the developers, they are running mpl after cloning from GitHub. So, if the idea is to `not store baseline images in the repo at all` as written above. Does this mean deletion of https://github.com/matplotlib/matplotlib/tree/master/lib/matplotlib/tests/baseline_images folder from release 3.3? I think we should do the deletion step after merging prs for #16477 and #11732. This will not cause much trouble for developers already working on the release 3.2 (when separate packaging changes and baseline images problem changes are complete in release 3.3).
 
I assume this deletion of baseline_images folder is meant for #16447, not for #11732. Correct?

For the developer, after upgrading from 3.2 to 3.3 release, developer has to install mpl-tests too as they don't have baseline_images folder right now. Correct?

Thanks for helping me out in better understanding of the problem and discussing the various use cases. I believe the project is complex in logical ability and challenging too. But your regular guidance is helping to understand completely new domain of knowledge, which I never looked in past. Thanks all :)

**Comment by SidharthBansal** [comment]
> The idea was that if someone downloads matplotlib from PyPI (either as a .tar.gz or as a .whl), they should get something self-contained from which it should be possible to run the test suite without any further internet access or git repo

I was reading https://pip.pypa.io/en/latest/user_guide/#installing-from-local-packages (linked from https://pip.pypa.io/en/latest/reference/pip_install/#caching). Can the developers use `pip install --no-index --find-links=DIR -r requirements.txt` command to create remaining baseline images?

**Comment by SidharthBansal** [comment]
Can we use [pytest-diff](https://pypi.org/project/pytest-diff/) for finding tests which are deleted or added across dev's master and last release?

**Comment by anntzer** [comment]
Sorry I forgot to reply to the previous message.  Feel free to reping as needed, at worst I'll say I don't have time to reply right now, or possibly someone else will reply ;)

> So after merging #11732, release 3.3 of matplotlib(which doesn't contain tests) will override the old package release 3.2 matplotlib(which contains tests). So, this upgrade will decrease storage for users(not developers) of mpl as now tests will be deleted from their computers. Is it right?
I believe users will be using commands like pip install --upgrade matplotlib for upgrades.

It will decrease storage for users, although some third party packagers already provide Matplotlib-wthout-test-data.  But even taking these into consideration, the nice thing is that after #11732, when someone installs Matplotlib, they'll know exactly what they are getting: Matplotlib without the test data.  Compare with today, where depending on where you install Matplotlib from (PyPI, conda, linux distro package, homebrew, etc.) you may or may not get test data.

> For the developers, they are running mpl after cloning from GitHub. So, if the idea is to not store baseline images in the repo at all as written above. Does this mean deletion of /lib/matplotlib/tests/baseline_images@master  folder from release 3.3? I think we should do the deletion step after merging prs for #16477 and #11732. This will not cause much trouble for developers already working on the release 3.2 (when separate packaging changes and baseline images problem changes are complete in release 3.3).  

Yes.  Though of course the deletion should only happen after the rest of the machinery is in place (in particular, the "add-new-baseline-image-by-running-test-suite-at-commit-that-introduced-it" machinery must be implemented).  Also we don't have to delete the baseline images directory *immediately*, we can leave it for the transition period.

> I assume this deletion of baseline_images folder is meant for #16447, not for #11732. Correct?

Yes.  #11732 just splits the artefacts (wheels) we produce into two, but they are made from the same repo.

> For the developer, after upgrading from 3.2 to 3.3 release, developer has to install mpl-tests too as they don't have baseline_images folder then. Correct?

Either that, *or* they should be able to regen themselves the baseline images locally by running the test suite on a known-good version (e.g. 3.3).  Again this ability needs to be present (I believe) to make addition-of-new-baselines-between-tags work.

> Thanks for helping me out in better understanding of the problem and discussing the various use cases. I believe the project is complex in logical ability and challenging too. But your regular guidance is helping to understand completely new domain of knowledge, which I never looked in past. Thanks all :)

I think you are doing very well in grasping all the parts of this project.

> I was reading pip.pypa.io/en/latest/user_guide/#installing-from-local-packages (linked from pip.pypa.io/en/latest/reference/pip_install/#caching). Can the developers use pip install --no-index --find-links=DIR -r requirements.txt command to create remaining baseline images?

If they are packaged, yes, but again (sorry I didn't reply to your previous message earlier), there are cases where they won't be packaged yet and they should be able to generate them locally.  On the other hand it is also true that we may just instead publish nightly wheels (there's already nightly.matplotlib.org but I don't know the details about that, perhaps @tacaswell can help?) for the test baselines, that could help too?  I haven't thought too much about it.

> Can we use pytest-diff for finding tests which are deleted or added across dev's master and last release?

I don't think that's what that plugin does?  On the other hand I think (untested...) it should be "easy" to find the diff by running something like `pytest --collect-only` and parsing the results, or more robustly via a pytest hook (https://docs.pytest.org/en/latest/writing_plugins.html#writing-hook-functions); e.g. use a custom run env where you have a pytets_collection_modifyitems (https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_collection_modifyitems) hook that just dumps the list of tests to a file and immediately exits.

**Comment by SidharthBansal** [comment]
Thanks for responses :-)

> Also we don't have to delete the baseline images directory immediately, we can leave it for the transition period.

So, are we will target deletion of test folder in 3.4 or onwards releases. We are keeping 3.3 for #16447 testing. OR do you mean that we will do it in September/October(that is after GSoC). I will love to contribute even after GSoC. 

Is there any release period followed? Like new release every quarter or month is compulsory?

**Comment by SidharthBansal** [comment]
So, #16447 and #11732 changes will go initially in the master branch. Then v3.3.x will be created and changes will be copied to that branch from master at the time of release. Please correct me if any other approach is followed here. 

I learnt from a few days ago that v3.2.2-doc is for docs as is independent of master changes. Will there be any changes related to documentation in that branch too? If we will be publishing it to the website, then there should be changes. But as you mentioned about the transition period. So, shall that documentation of website changes be at end of GSoC(/after transition period)?

I guess there should be changes in https://github.com/matplotlib/matplotlib/blob/master/doc/api/api_changes_3.3/development.rst. Is there any need for the creation of testing.rst? As test folder will be removed so removals.rst should also be changed. I don't think there should be any change in behaviour.rst. What do you suggest? Are there any other parts of documentation which need a change?

For mpl-test-data we will be using tags Or

**Comment by anntzer** [comment]
> So, #16447 and #11732 changes will go initially in the master branch. Then v3.3.x will be created and changes will be copied to that branch from master at the time of release. Please correct me if any other approach is followed here.

3.3 will be created a bit before the release but yes, things should go into master first and branched into 3.3.

> I learnt from a few days ago that v3.2.2-doc is for docs as is independent of master changes. Will there be any changes related to documentation in that branch too? If we will be publishing it to the website, then there should be changes. But as you mentioned about the transition period. So, shall that documentation of website changes be at end of GSoC(/after transition period)?

> I guess there should be changes in /doc/api/api_changes_3.3/development.rst@master . Is there any need for the creation of testing.rst? As test folder will be removed so removals.rst should also be changed. I don't think there should be any change in behaviour.rst. What do you suggest? Are there any other parts of documentation which need a change?

Let's not worry about the doc branch for now; these are mostly for backports anyways.  Having the right docs in the master branch should be good enough.

> For mpl-test-data we will be using tags Or

I guess these will reuse the same tags as the main wheels.

**Comment by SidharthBansal** [comment]
Please reply to whenever you get time https://github.com/matplotlib/matplotlib/issues/16447#issuecomment-615963608

Thanks

**Comment by anntzer** [comment]
Sorry I missed that.
The release schedule is pretty flexible (and we haven't been very good at staying on schedule recently anyways -- that's what you get when (nearly) everyone is just volunteering free time), 3.3 is probably going to be out soonish so likely this should be targeted for the 3.4 release cycle, but this shouldn't really change the gist of the work.  If you're willing to stay after GSOC ends that would of course be even better :)

**Comment by SidharthBansal** [comment]
Yeah, I will be after gsoc too.

**Comment by SidharthBansal** [comment]
Will it be better if we target #11732 in release 3.3 and #16447? This will help us in debugging the issues of #11732 in #16447.  What do you suggest?

**Comment by anntzer** [comment]
That's fine with me too.

**Comment by SidharthBansal** [comment]
> for CI we can rely on CI cache or stand up a store on github/DO/AWS/... managing that cache can be done via github actions

In case we will store the baseline images then we need to maintain it every time when we push a change related to the baseline images' tests. If we switch to CI cache so it will analogous to the local development, I guess.
Which one do you suggest? I think CI cache. But will be happy to learn about storage and GitHub actions too. :)

**Comment by anntzer** [comment]
CI cache seems fine.  That seems closest to being similar to what would happen on a dev machine.

**Comment by SidharthBansal** [comment]
I was a little busy submitting the changes for the journal last week. So, I couldn't focus much here. Sorry. 

If the developer's current baseline images is B and after `git pull` there are some insertions and deletions. I think the process should automatically generate new baseline images and delete the redundant baseline images. Correct me if I am wrong(or is there any need for developer to run any command. Running command doesn't make sense to me as it is frustrating to run a command each time after git pull)

I think the comparisons should be done on `git pull` instead of `git checkout` as it is less erroneous. Also, rebasing is considered by considering git pull. What do you suggest?

**Comment by anntzer** [comment]
I think doing work on git pull would be overly costly given that not everyone with a git repo will want to actually run tests.  You can instead do the work when someone invokes pytest (in one of the pytest startup hooks, e.g. pytest_configure).

**Comment by SidharthBansal** [comment]
Ok. I will check it out. Thanks

**Comment by SidharthBansal** [comment]
Copying from chat so that discussion is at one place
There is a cron job on https://github.com/MacPython/matplotlib-wheels that
builds the wheels and uploads them to https://nightly.matplotlib.org

**Comment by SidharthBansal** [comment]
> if a baseline image needs to change (we should still try to keep this as rare as possible), the commit message needs to include a special string e.g. [baseline-change]: path/to/image. The process described in 2) would also consult the git log to know if there's any images that need to be gen'd in such a way. Likewise, if a baseline image needs to be added for a new test, the commit message needs to be tagged accordingly. 

I think git tags stored whenever there is addition/deletion of the baseline images will ease the process by simply reading the git tags first followed by commit messages. This will reduce complexity. However, the main disadvantage is the developer needs to make git tag at each baseline image change. (might be little frustrating from developer's end but a good practice). So, do you suggest changes in git tags too when baseline images changes?

**Comment by SidharthBansal** [comment]
> Possibly, we could also require that the filename also be changed in that case so that even if you're not in a git repo, you at least don't end up using the wrong baseline image, and instead just get a skip.

Do we need to trigger the developer to run pytest so that they can be on the latest version of baseline images someway?
Considering the use case when there is addition

**Comment by anntzer** [comment]
> git tags

I guess the problem of git tags is that you can't PR one, so this would need maintainers to not only merge the PR but also remember to tag the commit.  And to read git tags you anyways need git, and once you have git present it's not too hard to find out the last commit where a file was modified (`git log -- path/to/file` -- I think keeping individual (empty) marker files for each baseline image is likely simplest here).

> Do we need to trigger the developer to run pytest so that they can be on the latest version of baseline images someway?
Considering the use case when there is addition

I *guess*(?) this only matters if there's two people working on the same baseline images at the same time?  That seems rather unlikely tbh.  Or did I miss something?

**Comment by SidharthBansal** [comment]
> I think keeping individual (empty) marker files for each baseline image is likely simplest here

Sorry. I am facing difficulty in understanding this.

We will have either insertion or deletions of files. So, do we still need to keep the deleted files appended with some suffix for recognization?

**Comment by SidharthBansal** [comment]
> I guess(?) this only matters if there's two people working on the same baseline images at the same time? That seems rather unlikely tbh.

Correct. It is rare but still what we need to do for that?

**Comment by anntzer** [comment]
> Sorry. I am facing difficulty in understanding this.
> We will have either insertion or deletions of files. So, do we still need to keep the deleted files appended with some suffix for recognization?

Sorry for being unclear.  At any commit there's a set of baseline images in use (test_foo/img0, test_bar/img1, ...).  I suggest keeping empty marker files (baseline_images/test_foo/img0, baseline_images/test_bar/img1, ...) corresponding to these.  Baseline images not in use anymore likewise have their marker files deleted.  Even if a marker file has been deleted, `git log -- path/to/file` will also give you the history of that file, i.e. when it was added or deleted.  Indeed, now that I think of it, this should be much more efficient than having tags for each change, as a tag wouldn't tell you what images changed (unless you encoded that in the tag name, but then you get enormous names if multiple files are changed at once, and you're basically just reinventing a filesystem in the git tag namespace...)

> Correct. It is rare but still what we need to do for that?

If we just use marker files with an incrementing counter, we'll get a conflict on merging the second marker file (I guess we may want e.g. to add a small random string in each marker file to prevent git from doing a trivial merge) and can resolve that as any normal merge conflict.

**Comment by SidharthBansal** [comment]
So, everytime a new baseline image is added, then there will be addition of baseline/test_x/img0(marker image) and test_x/img0(actual image in use). Deletion will have only deletion of baseline/test_x/img0(marker image). If this is the correct understanding, then marker files suppose to be empty files. Correct?

Also, consider a future use case, a lot of baseline images deleted and added. Then all deleted images' markups will take some space(although very small). I think that can be removed manually by maintainer after 5-10 releases.

**Comment by anntzer** [comment]
> So, everytime a new baseline image is added, then there will be addition of baseline/test_x/img0(marker image) and test_x/img0(actual image in use). Deletion will have only deletion of baseline/test_x/img0(marker image). If this is the correct understanding, then marker files suppose to be empty files. Correct?

Yes.

> Also, consider a future use case, a lot of baseline images deleted and added. Then all deleted images' markups will take some space(although very small). I think that can be removed manually by maintainer after 5-10 releases.

They will only be in the git history, which by definition is immutable, so they're going to be there forever anyways?

**Comment by SidharthBansal** [comment]
Agreed

**Comment by SidharthBansal** [comment]
Agreed, thanks. 

While a developer is raising a pr for deletion of test related to addition of baseline_images, should the developer also delete the corresponding markup files themself? OR we need to automate it somehow.
Similarly, for adding a test which produces an images, then developer should also add markup file for that image too.
So, for modification of same featured baseline images, do we consider versioning histogram1.0, historgram1.1 like that?

**Comment by SidharthBansal** [comment]
I think https://github.com/matplotlib/matplotlib/tree/master/lib/matplotlib/tests/baseline_images can be changed into markup files as last step of #16447.

**Comment by anntzer** [comment]
I think the developer should do it (or rather we should have tooling to do it, see e.g. how tools/triage_tests.py can currently be used for copy new baselines).
Versioning can just use an incrementing counter, no need to overdo it for now.

**Comment by SidharthBansal** [comment]
> I think the developer should do it (or rather we should have tooling to do it, see e.g. how tools/triage_tests.py can currently be used for copy new baselines).

I think that can be a follow up PR after #16477

**Comment by anntzer** [comment]
yes

**Comment by SidharthBansal** [comment]
I think Work Break Down Structure could be:
1. Completion of #11732 
2. Generation of known baseline images
3. Caching for grabbing baseline images
4. Addition/Deletion of baseline_images from git commit hash and git log
5. Travis related work(analogous to above). 
6. Optimisation like https://github.com/matplotlib/matplotlib/issues/16447#issuecomment-623317655
Or shall I start with some other piece of implementation?
What do you suggest? 
Do we need any other changes in CI except Travis?

**Comment by anntzer** [comment]
Sorry for the delay; here's a slightly more detailed task list:

-----

First, #11732 (generation of test-only wheels) should be completed.  This PR splits the PyPI sdists/wheels in (a) one artefact that does not contain baseline images, and (b) one that contains only the baseline images; (a) is what one gets when running `pip install matplotlib` and (b) is what one gets when running `pip install matplotlib-test-data` (where `matplotlib-test-data` (name up to bikeshedding) is a new project that we need to register on PyPI); each release of `matplotlib-test-data` should install_requires *exactly* the corresponding version of `matplotlib`.  Preferably the same split should be done on the conda packages, but this can wait a bit.  This would be generally useful even regardless of the rest of the project, as this ensures that we are in a "known state" wrt. baseline images, whereas currently it is never clear whether baseline images are present when installing matplotlib.

IIRC (that PR was a while ago) #11732 was mostly good, but it may be needed to check how it plays with editable installs.  In particular, what if matplotlib is installed normally and the test data editably, or vice-versa?  We may not have to support all combos, but should at least "error out" cleanly.  It may also help to first move the baseline images out of the main source tree (i.e. under lib/matplotlib/...) to some other place still in the git repo, to avoid confusion with editable installs.

Once #11732 is implemented (likely after 3.3 is released), we switch to having two such wheels and can use e.g. nightly.matplotlib.org as "test PyPI server" i.e. if one installs a prerelease of matplotlib from nightly.matplotlib.org one should get the testless version and one should be able to install matplotlib-test-data from it as well.

The next feature to implement would then be being able to run the tests *without* matplotlib-test-data being installed (by fully regen'ing baseline images from a previous commit), and then comparing the current commit against it.

Then (things become more distant now) machinery to add "new" baseline images (i.e. just marker files saying "a new baseline image was added here"), together with machinery to make PRs still reviewable in that case (e.g., perhaps a bot that posts the images to the PR thread).

Only after all these tasks are implemented, we can then delete the old baselines from the repo, replacing them all by the marker files.

-----

Also, I'm not sure a single issue thread is really the best place to keep track of the progress of this, but I'm not sure I have much better ideas (e.g. github discussions would not be public, which is less than great; perhaps discourse? @story645 can we have a category there? or any other suggestions?)

**Comment by story645** [comment]
I've half heartedly used GitHub projects for gsoc and think it's not a bad idea if this gets broken down over multiple PRs. 

GitHub discussions can be public far as I know, but @tacaswell /@dopplershift /@efiring /@mdboom are the ones who can turn it on.

 Also fine w/ making a GSOC section on discourse under the development category.

**Comment by tacaswell** [comment]
I can not figure out how to turn on "discussions".  :disappointed:

**Comment by dopplershift** [comment]
To be clear, GitHub "Team Discussions" are something we have access to right now, but these are not public.

"GitHub Discussions" are the new message-board-like feature they showed this week, and will be publicly accessible, but will be available to public repos ["soon"](https://github.blog/2020-05-06-new-from-satellite-2020-github-codespaces-github-discussions-securing-code-in-private-repositories-and-more/#discussions).

**Comment by anntzer** [comment]
We're going to use github projects + gitter for now.

**Comment by SidharthBansal** [comment]
Here is the link https://gitter.im/matplotlib/GSOC so that anyone who wishes to join can join :)

**Comment by SidharthBansal** [comment]
**User Stories**
1) Developer D installs the mpl repository on their local computers. Thereafter, they started writing tests. So, they will need to install the baseline images. They installed the baseline images using the matplotlib_baseline_images package. Thereafter, the added test named `new_test`. This `new_test` compares against a new baseline image. That baseline image is not present initially. What should be done? Where will be the images stored?
2) Developer D installs the mpl repository on their local computers. Thereafter, they started writing tests. So, they will need to install the baseline images. They installed the baseline images using the matplotlib_baseline_images package. Thereafter, they deleted the matplotlib_basleine images package(`git clean -xfd`). Thereafter, the added test named `new_test`. This `new_test` compares against a new baseline image. That baseline image is not present initially.  What should be done?
3) Similarly for the generation of the baseline images through command line
4) Where will the `generate_images` create the missing images if there is no baseline images directory. Should that be first copied from the mpl_cache?

**Comment by SidharthBansal** [comment]
**Proposal for git diff**
_________________________

   1) Tell travis to ignore test failure on new image
   2) Create diff for existing image on master and post to github as comment on PR
   3) Tell CI to generate new baseline on merge-to-master

Whenever a developer modifies a baseline image then they create a record in the modifiedBaselineImageTest file 
The record will be of the syntax
`Type FileName:TestName (To) (FileName:TestName)`There will be one record per line.sac nType: Added, Modified, Deleted
In case of the `ADD` type, then we will run the new testfile - testname
In case of `DELETE` noops
In case of `MODIFY` generate the image for testfile - testname at both master and the author changes. Then show the reviewer both the images.

**Comment by tacaswell** [comment]
Instead of writing a custom format, can we embed this information in json or yaml?

**Comment by SidharthBansal** [comment]
Ya, we can use either of the two.
In case any file name was changed A -> B then B -> C, then A doesn't exist. For that reason we can have a check if the first file name exists then only we will do further processing

**Comment by SidharthBansal** [comment]
So, we can have json as 
```
{
"ADD": {
     "fileName": "testName"
}
"DELETE": {
     "fileName1": "testName"
     "fileName"2: "testName"
}
}
```

I think if developer will keep only the files/tests changed by them in this file, then this will be easier to parse  In case there are multiple possible such scenarios then it will take a lot of time to process redundant cases.

Also, I don't think we need to maintain this file on the master. What do you think?

**Comment by jklymak** [comment]
Coming back to this, my vote would be a separate repository for the images and folks who don't want to pull down the whole tree can set the depth to something shallow.  The only drawback in my mind is a little more test developer overhead, but "be sure to check in the images and open a related PR" seems easier to me than any of the instructions required for the solutions described above.

**Comment by dopplershift** [comment]
So an additional problem of using the "ancient" Freetype 2.6.1 is that it does not build on the mac-arm64 architecture, which makes it impossible to successfully run on that architecture. Before long, that's going to be the dominant architecture for the macOS platform. So are we committed to going in this direction? Or are we going to bump to a modern Freetype? Or...?

**Comment by QuLogic** [comment]
But we build wheels for arm64 and universal2 with bundled FreeType 2.6.1 already?

**Comment by dopplershift** [comment]
@QuLogic Apparently I missed that. Any idea if there was some magic needed? My recollection the last time I tried to get 2.6.1 to build on macos-arm64 (for conda-forge) was that it basically said "I have no clue what you're trying to build on".

**Comment by QuLogic** [comment]
Apparently nothing special: #20970

**Comment by dstansby** [comment]
Since this seems to be topical at the moment (https://github.com/matplotlib/matplotlib/pull/29816), I thought I would share the process used in https://github.com/astropy/astropy and https://github.com/sunpy/sunpy/ for generating/updating test images. I was partially responsible for building the pipeline (I think... it's been a while), and it works quite well.

The rough idea is to:
- Store hashes of generated figures in the main repository
- Store test images in a secondary repository
- When running a figure test, generate a hash of the figure
- Compare the hash against the expected hash in the main repository
- If the hashes differ, then:

1. Download the test image from the secondary repository
2. Use this to generate a visual diff for local comparison
3. Also generate a visual diff on CI for reviewers to inspect (see [here](https://output.circle-artifacts.com/output/job/ad7fb550-a191-44f8-99e7-b2c2ed370a2b/artifacts/0/.tmp/py312-figure/figure_test_images/fig_comparison.html) for an example of a HTML page generated for a sunpy commit)
4. If the difference is expected, commit the changed hash and merge to `main`
5. On merges to `main`, if any of the hashes have changed, GitHub actions re-generates the test image and pushes the updated image to the secondary repository

It would need a decent chunk of time to implement (like any solution to this problem I guess), but I think it works well from both a developer and contributor point of view, and it's been tried and tested for a couple of years in other projects.

**Comment by timhoffm** [comment]
Sounds interesting. I assume the hashes imply exact identity. Do you have the case that images change minimally within tolerances so that the reference image stays the same but hashes have to be regenerated?

Since #29816 requires a solution soonish, would a step-wise approach be feasible; i.e. flip out the reference images for hashes in the main library so that #29816 can update to new hashes and be merged. For a transition period we would then only have the hashes as reference and have to resolve any deviations manually. The image diff infrastructure could be build up in parallel in the background. Worst-case would be that we see that the approach does not work and we have to revert the #29816 variant with hashes and instead go with #29816 based on new images.

**Comment by dstansby** [comment]
> Do you have the case that images change minimally within tolerances so that the reference image stays the same but hashes have to be regenerated?

As far as I remember `sunpy` doesn't use tolerances, so if the bytes change the image is just regenerated. I guess with the images in a separate repo there is less 'cost' to uploading a new image, even if the pixels haven't changed. And it's on a slightly smaller scale - `sunpy` has 82 images, versus the hundreds(?) here.

As well as being used to avoid updating a test image, are tolerances also used to 'hide' differences in the images generated by different environments (OS, Python version, dependency versions etc.)? In that case I'm not sure off the top of my head how a 'hash based' solution could work. Perhaps one could store multiple hashes for each figure test, and as long as the generated hash matches one of the expected hashes the test passes? I'm not sure in that case you would know which of the expected hashes to updated when the figure changes though.

**Comment by greglucas** [comment]
@bjlittle has proposed using perceptual image hashes in pytest-mpl: https://github.com/matplotlib/pytest-mpl/issues/146 That seems like it would be nice here as well and a tolerance could be added to the hash distance even.

One thing that isn't clear to me is whether I _need_ internet access to do these things? Is there a local caching mechanism where I can bring the images down locally and work/compare without needing to retrieve the remote file all the time to see the difference (example if I'm on a plane, but I want to be able to compare actual image differences while testing not just hashes).

**Comment by greglucas** [comment]
> @greglucas Do you know what that looks like?

No. I have not played around with these things at all which is why I looped Bill in here because I believe he has :) Maybe @rcomer knows if they are using it on Iris too?

Maybe there is a way for us to figure out what the tolerance was for an entirely missing glyph and set the threshold just below that. That would theoretically give us multiple minor version updates of freetype with minor kerning adjustments and only once enough of the kerning changes all add up does it go over that threshold.

I am probably not the one to explicitly ask about the tolerances as I am more OK with looser tolerances. One idea is that we could try to make a distinction between categories of tests, freetype-specific / font-specific where I do care about tolerances, and plots where I am less concerned about minor glyph kerning and care more about the placement of a scatter marker in the right spot.

**Comment by dstansby** [comment]
> One thing that isn't clear to me is whether I need internet access to do these things?

If we store the test images in a repository, you could either point the location of the baseline images to be that remote repository, or you could clone it (before starting your internet free time) and point to the local checkout as the location of the baseline images.

> That would theoretically give us multiple minor version updates of freetype

The idea of using hashes is that updating the images is (almost) zero cost in terms of effort and main repository size, so I would not worry too much about having to update all the images for different versions of freetype to start with. I think to keep things simple here we should try out storing the images in a separate repository and using hashes - if that works, we could then explore whether there's still a need to reduce the frequency that we're updating test images (presumably to keep the size of the test image respository low?)

**Comment by dstansby** [comment]
Ahh I have just understood the motivation for using a perceptual difference instead of a hash - it would allow us to replicate the current behaviour of using tolerances in the tests, where using a cryptographic hash doesn't.

**Comment by tacaswell** [comment]
> I think our somewhat draconian tolerances arose because a missing glyph could pass looser tolerances.

Yes, we were using `\cdot` (one dot in the center) instead of `\ddots` (three dots diagonal) or something similar.  Because the contrast is black/white for text even minor shifts give huge RMS and the number of pixels "wrong" for dropping some dots is less than from shifting a lowercase `l` by 2 pixels.  I do not think there exists a tolerance that will reliably catch that sort of error but will ignore the sort of version-to-version drift we see in freetype.

----

https://github.com/tacaswell/mpl-imageset-demo and #25734 are working but lack documentation and sorting out how to make the required caching work on CI.

**Comment by tacaswell** [comment]
I also have general concerns about updating images too often and view it as a high cost (even if they are in a second repo) or high risk.  We have a lot of test images and if we are on the regular regenerating most of them, it will either require a lot of manual work to validate that the "small" changes are in fact small-from-a-communication point of view or we are going to be lax and end up committing very wrong baseline images again.  

An additional concern with hashes is that we put the version + date in the files as metadata so we can not hash the files.  Additionally for the vector backends we want the tests to be insensitive to things like the order of things in the file, local identifiers, and other non-meaningful changes to the files (e.g. dropping the number of digits we write out for floats).   Currently the way we do this is to checking the svg/pdf/ps and then at test time use the same viewer to rasterize the files and compare the pixels of the rasterization.  If we were to track the hash of that rasterization then we would add a dependency on the version of the viewer we used (and any platform dependent variation they have).

-----

> One idea is that we could try to make a distinction between categories of tests, freetype-specific / font-specific where I do care about tolerances, and plots where I am less concerned about minor glyph kerning and care more about the placement of a scatter marker in the right spot.

If we are going to go through and re-generate a lot of test images it should be coupled with an audit of if that image _actually_ needs text in it and switching them to the new defaults.

**Comment by timhoffm** [comment]
> An additional concern with hashes is that we put the version + date in the files as metadata so we can not hash the files.

That's not a fundamental issue. Worst-case, we hash the RGB image values.

**Comment by rcomer** [comment]
Iris’s decision to use perceptual image hashes was before I got involved, so I’m not really up on the pros and cons.  I guess @pelson would have been involved back then so might have an opinion whether they’d be appropriate here.

**Comment by timhoffm** [comment]
For any kind of hash, we can use them as a pre-test for an image test if the hash is reasonably stable. If the hash is "known-good", we are done, if not we have to fall back to the image test (possibly downloading the reference). If the image test is successful, we add it to the "known-good" hashes.

This works as long as long as there is a limited number of image variants. It wouldn't work if there's a small randomness in every generated image.

**Comment by bjlittle** [comment]
@greglucas Thanks for the nudge and looping me in ...

I'm still very keen to close https://github.com/matplotlib/pytest-mpl/issues/146. The original pull-request https://github.com/matplotlib/pytest-mpl/pull/150 unfortunately stalled as I needed some `pytest-mpl` core dev support to help me cross the line. Also my proposal was a major breaking change in behaviour and some big decisions still required to be agreed and ironed out. Anyways, I kinda lost momentum, life got in the way (as it does), and I closed the PR to keep things honest. That's all totally on me btw.

In terms of https://github.com/SciTools/iris, it was my decision to implement and introduce perceptual image hashing, and thankfully it's really worked out for us, in that it's given us the stability and control that we required. We've been happily using it for many years now to perform our mpl image testing.

Anyways, I'd love to blow the dust off https://github.com/matplotlib/pytest-mpl/pull/150 and resurrect it.

Perhaps what would give me more confidence and conviction to actually follow through and deliver would be if I could work with someone from the matplotlib dev community? Someone who could be a technical point of contact so that we could kick the tyres of perceptual image hashing to determine whether it addresses some/all/none of your issues. That would really give me the confidence to push the breaking change to `pytest-mpl` and more importantly it might help mitigate risk on your side, give you guys clarity on the impact and implications, and ultimately help decide whether this might be useful to you or not.

My proposal would be to work with @rcomer, who's also at the Met Office, and logistically that would make a lot of sense. I know this suggestion has come pretty much from left field, so have a think and let me know what you guys decide. In the meantime I'm going to tentatively see whether it's feasible to secure the time and space to do this work during my 9-to-5; I know of several groups at the Met Office that would like to see this happen, so that gives me some leverage to sell the idea.

Anyways, just a thought ...

**Comment by tacaswell** [comment]
> Is there not a dependency on the version of the viewer used anyway to do the rasterization?

We run the rasterization of both the baseline and the machine running the test with (presumably) the same version of the viewer so variation between how the veiwers render them should was out.

**Comment by timhoffm** [comment]
@dstansby FYI https://github.com/matplotlib/pytest-mpl/pull/150 has been superseeded by  https://github.com/matplotlib/pytest-mpl/pull/259

**Comment by cvanelteren** [comment]
@timhoffm no review is provided yet; does this mean this is in the scope of mpl?

**Comment by jklymak** [comment]
I played with Claude Sonnet 4.5 on this problem, as I admit I found it pretty hard to follow the ins and outs without more guidance.  

It seems possible to do a lot of things I'd not have thought of.. Overall, the basic (AI-generated) summary is:

> Hash-Based Baseline Image Proposal Summary:
>
> Core Approach: Replace storing 2,330 baseline images (40MB + 507MB Git history) with a lightweight `baseline_hashes.json` file (~200KB) containing perceptual hashes for each test's expected output. Actual baseline images are stored separately (GitHub release artifacts or dedicated repo) and downloaded on-demand only during test failures for visual inspection. This prevents repository bloat while maintaining full test coverage—if hash comparison passes, the test passes without downloading any images; only mismatches trigger image downloads for debugging.
>
> Platform Handling & Workflow: Perceptual hashes support configurable Hamming distance tolerance (default: 5 bits) to handle minor platform rendering differences automatically (antialiasing, slight font variations). For legitimate platform-specific rendering (macOS vs Windows), the JSON stores multiple valid hash variants per test. Contributors run pytest --accept-new-baseline locally to update the hash file; CI automatically uploads actual images to storage, verifies hash matches, and posts images for reviewer inspection. This eliminates manual image uploads, requires only single-PR contributions, and maintains visual review capability while drastically reducing repository size.

This all sounds pretty reasonable to me.  However, it definitely requires pytest _and_ CI infrastructure to make work.  I particularly like the idea of the CI posting image diffs to Github for visual acceptance.  This still means that there is image storage somewhere, but the user never has to download them all going forward, only the ones for tests that they are running.  

I think this can be implemented on an incremental basis, in that we could start the hash file, and the "hashed_images" repository while still allowing the old method, and convert old files as needed.  

Details at at https://gist.github.com/jklymak/cb5ace18dd4d9b2359ec5f9aca1b2fda
